### DISCUSSION TITLE
Building in-cab control for multiple OWLs

### DISCUSSION THREAD

**Discussion by geezacoleman** [geezacoleman/OpenWeedLocator#65] [Ideas]
It's be a while coming (first raised in Issue #5 ) but I've started work on an MQTT protocol for communication between multiple Raspberry Pis mounted on a boom.

The architecture would be:
- One Raspberry Pi acting as a MQTT broker and also running an OWL GUI that subscribes to relevant topics
- The Pis on the boom would publish information on: 
  - each detection/weed location (assuming GPS integration)
  - status - i.e. any errors or otherwise
  - subscribe to an 'operational' topic for stop/start/blanket spray etc. instructions
  - subscribe to 'parameters' topic for sensitivity/changes in algorithm
- cab-mounted Pi would then display when nozzles are on and allow user to interact and adjust performance remotely with a screen/button interface.

Have started on this in the [mqtt-comms](https://github.com/geezacoleman/OpenWeedLocator/tree/mqtt-comms) branch with a basic PyQt5 UI. All ideas/contributions/suggestions are welcome!

**Comment by Minorplanets76** [comment]
AgOpenGPS has migrated to UDP via ethernet with messages being sent via PGN's.  The upgrade has done wonders for system response and reliability.  CANBUS would be another possibility given its use in many tractor and spray technologies.  There is a branch looking at rate control as well, and ESP based wifi which may provide some inspiration.

**Reply by geezacoleman** [reply]
Is it worth pursuing the AgOpenGPS approach over MQTT? I'm not super familiar with UDP/PGN. What were they running before?

CANBUS also makes a lot of sense given how widely used it is - lots of existing displays use CANBUS for in-cab control from what I've read. Only issue is the extra hardware required on each Pi - MQTT is attractive in that sense, it really only needs the ethernet cable. Wi-fi makes me a little concerned given all the electromagnetic noise around a large bit of machinery but running cables is never too much fun.   What do you envisage as an optimal system?

**Reply by Minorplanets76** [reply]
They're probably pretty similar really. A message is broadcast on particular ports and components look for the messages they're looking at. There's been plenty of discussion and I don't really have the expertise to comment too much, but from the AgOpenGPS experience, it was huge leap in reliability moving away from USB connections and Arduinos talking over i2c or serial. That said I've also learned a lot about wiring in that time so less dodgy twisted together wires wrapped in electrical tape shorting out. 

Spray systems seem to be using Isobus at the high end. That's a whole other system I don't understand 😉

**Reply by plains203** [reply]
MQTT is a good option for comms. Means we can integrate it with a bunch of others things easily.

**Reply by geezacoleman** [reply]
Any suggestions on topic/message standards to follow that would make it easier to integrate? 

MQTT seems the lowest overhead (no extra parts/can just use ethernet cabling and switches). Also had the suggestion to use Fast API to develop a web interface using ethernet + mqtt I suppose. Comms protocols are well beyond what I'm familiar with in computer vision!

**Reply by plains203** [reply]
https://oxrs.io/docs/firmware/touch-panel-esp32.html checkout this project for an option for a control panel. I would use the tasmota style stat/ cmnd/ Tele/ style topic structure

**Reply by Minorplanets76** [reply]
I suppose it depends on what you need to communicate. In its simplest form just enable/disable.  I'm sure you would want the trigger events sent back from each unit if only for the visual confirmation that it is working. 

At this stage, using Ethernet makes a lot of sense. That would allow to ssh in to individual units, push updates etc. 

There would be no reason why the Pi's couldn't do some system monitoring like pressure or solenoid health as well. 

A synchronised time seems important to if logging data. That's probably a conversation for the GPS interface.

**Reply by EchelonMirror** [reply]
Ethernet can get messy on larger booms where are you are considering potentially 30 ethernet cables on a 30m boom plus the fact that you will need to have a switch to handle all that traffic. 

Wifi is a fun solution with low bandwidth requirements and can scale up easily. https://wiki.teltonika-networks.com/view/RUT_Routers are decently priced with good UI

**Reply by geezacoleman** [reply]
Those RUT routers look pretty handy. How reliable have you found wifi for industrial settings? That's my main concern with it, but it's probably misguided! Which model RUT have you used before?

For the ethernet idea, you would have one switch per boom section and then one central switch (like the diagram below) to connect everything to an in cab controller. The controller could be running a FastAPI web app with mqtt (for commands/status/other data) + ethernet (for video feeds) running over the LAN. This interface would allow for all range of hardware to set up the network I imagine, but switches + cables seems the most straightforward and low cost at least for smaller systems?

![image](https://github.com/geezacoleman/OpenWeedLocator/assets/51358498/1ee151a5-2241-481d-9ecb-0118f278062b)

Was also thinking that the core switch/in-cab one could have a wireless access point so you could use a tablet/wireless controller to manage everything

**Reply by plains203** [reply]
Don't forget you can use the raspberry pi as a handy wifi router as well. Could dedicate one to running the interface and use that as a wireless router for the others.

**Reply by Minorplanets76** [reply]
Similar concerns were raised with AOG which is why it went down the wired ethernet route.  Plenty of spirited discussion, but in the end the bloke coding it needs to do it the way it suits them best.  Some have used WiFi (Esp32's) had no issues and no lag.  WiFi cameras are becoming commonplace.   In this application, would it matter?Each OWL unit is working on its own with the critical camera to nozzle timing being handled locally.  I could see it becoming a big deal if you had a camera out front talking to a separate unit responsible for triggering solenoid.  Maybe if synchronisation was important for mapping or something?

I have a Pi4 running my WiFi in my workshop which it's been doing reliably since before the Pi supply crisis and long enough ago that I've forgotten how I set it up.   I considered pinching it for OWL but was too frightened I would wreck the WiFi.  RasAP was the app and was pretty easy to configure.  Mine's configured as an access point so everything over there is its own  network, but I'm pretty sure it can be configured like a network extender too.  Coverage is probably about 20m or so.

**Reply by EchelonMirror** [reply]
I moved to Wifi mostly for scalability, and some future plans with error detection, calibration via tablet etc. Ethernet switches that are IP67 compliant are prohibitively expensive. For small scale systems, an 8-12 port switch will do the job if you put it in a housing that is compliant.

For large scale systems, I've had issues with EMI caused from tractors and large noisy power sources. geezacoleman, someone in your fam might be of help here to avoid some of the issues of larger hardware.

**Reply by sunlite71** [reply]
Dear Minorplanets76, Just curious have you tested Owl with Esp32 Cam?
Your earliest reply will be highly appreciated :-)

thanks

**Reply by ahogan27** [reply]
Where did this line of thought end up @geezacoleman? To Wifi or not to Wifi?

**Reply by geezacoleman** [reply]
Though fairly uninformed about wifi, I'm still of the belief that the hard wiring of modules is more reliable. Spoke to DeepAgro (Argentine spot spraying company) and they mentioned wifi had caused some trouble. Ethernet would allow video streaming and MQTT commands. But, a CAN system seems to be most common, particularly if we want to communicate with third part screens or robots (e.g. Agrointelli or SwarmFarm), I have also heard bad things about CAN and reliability.

So I think end result will be ethernet and MQTT - it's most future proof, easily done and could potentially be switched with wifi in the future if needed.

**Reply by EchelonMirror** [reply]
CANBUS is ancient and unreliable. ISOBUS is heading in the same direction since the standard was developed in the 90s.

Ethernet is hard to beat on reliability, though a lot of people will complain about wiring up extensive amounts of looms and the question of where you put the switch. Since it's going to need power and easy access for cables, in addition to IP64+ casing. 

Wifi can work if you reduce the number of units you have on the boom. I've been testing with a number of routers from Teltonika (RUT655) on a 36m boom and it's mostly stood up to scrutiny - I must note this is with 2-3m spacing with custom antennas and an Azurewave AW-CB375NF. Key is to make sure your devices have a rough LOS (line-of-sight) to your router. I'm using MQTT commands on that system. 

Basically, there is a core that acts a web server and MQTT broker. The goal was to be able to use a web browser as the control system through a cheap android tablet. Being purely software, it allows you to roll up new features with a single software update of the core. 

Do the folks at DeepAgro use real-time streaming?
